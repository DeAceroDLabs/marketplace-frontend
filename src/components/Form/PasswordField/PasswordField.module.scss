@use "styles/colors";
@use "styles/typography";

.container {
    display: flex;
    flex-direction: column;
    margin-top: 18px;
}

.input {
    height: 40px;
    border: solid 1px #666666;
    border-radius: 4px;
    margin-top: 8px;
    padding: 9px 16px;
}

input[type=password]::-ms-reveal,
input[type=password]::-ms-clear {
    display: none;
}

.error-text {
    margin-top: 10px;
    color: colors.$error;
    font-size: typography.$body-s;
}

.input-error {
    border: solid 1px colors.$error;
}

.input-error:focus-within {
    outline: none !important;
    border: solid 2px colors.$error;
}

.input-error:focus-within {
    outline: none !important;
    border: solid 2px colors.$error;
}

.visibility {
    float: inline-end;
    margin-left: 82%;
    margin-top: -30px;
    position: relative;
    z-index: 2;
}


@media(min-width: 450px) {

    .visibility {
        margin-left: 92%;
    }

}