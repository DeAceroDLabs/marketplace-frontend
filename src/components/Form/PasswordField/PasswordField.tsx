import { OptionsField } from "forms/form.types";
import { useState, useEffect } from "react";
import { useFormContext } from "react-hook-form";
import styles from "./PasswordField.module.scss";
import validator from "validator";
import VisibilityIcon from "@mui/icons-material/Visibility";
import Button from "components/common/Button";

const PasswordField: React.FC<OptionsField> = ({
  name,
  value,
  placeholder,
  label,
  required,
  type,
  disabled,
  errorMessage = "",
  needsValidationFrom = "",
}) => {
  const [error, setError] = useState("");
  const methods = useFormContext();
  const [currentValue, setCurrentValue] = useState(value);
  const [visible, setVisible] = useState(false);

  const validatePassword = (password: string) => {
    if (
      validator.isStrongPassword(password, {
        minLength: 8,
        minLowercase: 1,
        minUppercase: 1,
        minNumbers: 1,
        minSymbols: 1,
      })
    ) {
      setError("");
      return true;
    }
    setError(errorMessage);
    return false;
  };

  useEffect(() => {
    methods.clearErrors();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [error]);

  const confirmPassword = (confirmPassword: string) => {
    const passwordValue = methods.getValues()[needsValidationFrom];
    let confirmPasswordValue = methods.getValues()[name];
    confirmPasswordValue = confirmPassword;
    if (confirmPasswordValue === passwordValue) {
      setError("");
      return true;
    }
    setError(errorMessage);
    return false;
  };

  const activeError = error !== "";
  const emptyFieldWhenRequired =
    methods.formState.errors[name] &&
    methods.formState.errors[name].type === "required";
  const valueNotChanged = currentValue === value;

  const errorStyle =
    activeError || (emptyFieldWhenRequired && valueNotChanged)
      ? "input-error"
      : "";

  const requiredMessage = emptyFieldWhenRequired && valueNotChanged && (
    <span className={styles["error-text"]}>Este campo es requerido</span>
  );

  activeError &&
    (methods.formState.errors[name] = { type: "validation", message: error });

  return (
    <div className={styles.container}>
      <label>{label}</label>
      <input
        {...methods.register(name, {
          value,
          required,
          validate: (value) => {
            let isValid = validatePassword(value);
            if (needsValidationFrom) {
              isValid = confirmPassword(value);
            }
            return isValid;
          },
        })}
        className={`${styles.input} ${styles[errorStyle]}`}
        type={!visible ? type : "text"}
        placeholder={placeholder}
        disabled={disabled}
        onPaste={(e) => {
          e.preventDefault();
          return false;
        }}
        onCopy={(e) => {
          e.preventDefault();
          return false;
        }}
        onChange={(e) => {
          const value = e.target.value;
          setCurrentValue(value);
          validatePassword(value);
          needsValidationFrom && confirmPassword(value);
        }}
      />
      <div className={styles.visibility}>
        <Button
          color="transparent-grey"
          action={(e) => {
            setVisible(!visible);
            e.preventDefault();
          }}
        >
          <VisibilityIcon />
        </Button>
      </div>
      {requiredMessage}
      {error !== "" && <div className={styles["error-text"]}> {error}</div>}
    </div>
  );
};

export default PasswordField;
