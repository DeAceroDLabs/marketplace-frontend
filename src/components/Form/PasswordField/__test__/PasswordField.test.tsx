import { render } from "@testing-library/react";
import { Option } from "forms/form.types";
import PasswordField from "../PasswordField";
import { fireEvent, screen, waitFor } from "@testing-library/react";

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  useFormContext: () => ({
    register: () => jest.fn(),
    handleSubmit: () => jest.fn(),
    errors: () => jest.fn(),
    formState: { errors: {} },
  }),
}));

describe("PasswordField", () => {
  const setup = (options: Option[]) => {
    return render(
      <PasswordField
        label={"Mock PasswordField"}
        required={false}
        placeholder={"password"}
        type={"email"}
        value={"mock value"}
        name={"password"}
        errorMessage={"mock error message"}
      />
    );
  };

  it("renders PasswordField with no options without problem", () => {
    const view = setup([]);
    expect(view).toMatchSnapshot();
  });

  it("simulates change on input", async () => {
    const view = setup([]);
    const input = screen.getByPlaceholderText("password");
    fireEvent.change(input, {
      target: { value: "password" },
    });
    await waitFor(() => {
        expect((input as HTMLInputElement).value).toBe("password");
    });
  });

  it("simulates change on input with error", async () => {
    const view = setup([]);
    const input = screen.getByPlaceholderText("password");
    fireEvent.change(input, {
      target: { value: "password" },
    });
    await waitFor(() => {
        expect(screen.getByText("mock error message")).toBeInTheDocument();
      });
  });

  const setupWithValidation = (options: Option[]) => {
    return render(
      <PasswordField
        label={"Mock PasswordField"}
        required={false}
        placeholder={"password"}
        type={"email"}
        value={"mock value"}
        name={"password"}
        errorMessage={"mock error message"}
        needsValidationFrom= {"password"}
      />
    );
  };

  it("simulates change on input", async () => {
    const view = setupWithValidation([]);
    const input = screen.getByPlaceholderText("password");
    fireEvent.change(input, {
      target: { value: "password" },
    });
    await waitFor(() => {
        expect((input as HTMLInputElement).value).toBe("password");
    });
  });
  
});
