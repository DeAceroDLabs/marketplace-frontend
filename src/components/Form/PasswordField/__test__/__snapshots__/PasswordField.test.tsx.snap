// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField renders PasswordField without problem 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <label>
          Mock PasswordField
        </label>
        <input
          class="input "
          mock="[object Object]"
          placeholder="password"
          type="password"
        />
        <div
          class="visibility"
        >
          <button
            class="button transparent-grey activeClass"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="VisibilityIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <label>
        Mock PasswordField
      </label>
      <input
        class="input "
        mock="[object Object]"
        placeholder="password"
        type="password"
      />
      <div
        class="visibility"
      >
        <button
          class="button transparent-grey activeClass"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
